File: src/api/dotnet/OnClause.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    OnClause.cs

Abstract:

    Callback on clause inferences
    
Author:

    Nikolaj Bjorner (nbjorner) 2022-10-19
    
Notes:


--*/

using System;
using System.Diagnostics;
using System.Linq;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Microsoft.Z3
{

    using Z3_context = System.IntPtr;
    using Z3_solver = System.IntPtr;
    using voidp = System.IntPtr;
    using uintp = System.IntPtr;
    using Z3_ast = System.IntPtr;
    using Z3_ast_vector = System.IntPtr;


    /// <summary>
    /// OnClause - clause inference callback
    /// </summary>        
    public class OnClause : IDisposable
    {
        /// <summary>
        /// Delegate type for when clauses are inferred.
	/// An inference is a pair comprising of
        /// - a proof (hint). A partial (or comprehensive) derivation justifying the inference.
	/// - a clause (vector of literals)	
        /// The life-time of the proof hint and clause vector is limited to the scope of the callback.
        /// should the callback want to store hints or clauses it will need to call Dup on the hints
        /// and/or extract literals from the clause, respectively.
        /// </summary>                
        public delegate void OnClauseEh(Expr proof_hint, ASTVector clause);

        
        // access managed objects through a static array.
        // thread safety is ignored for now.
        GCHandle gch;
        Solver solver;
        Context ctx;
	OnClauseEh on_clause;

        Native.Z3_on_clause_eh on_clause_eh;

	static void _on_clause(voidp ctx, Z3_ast _proof_hint, uint n, uint[] deps, Z3_ast_vector _clause) 
        {
             var onc = (OnClause)GCHandle.FromIntPtr(ctx).Target;
             using var proof_hint = Expr.Create(onc.ctx, _proof_hint);
             using var clause = new ASTVector(onc.ctx, _clause);
             onc.on_clause(proof_hint, clause);
	}

        /// <summary>
        /// OnClause constructor
        /// </summary>        
        public OnClause(Solver s, OnClauseEh onc)
        {
            gch = GCHandle.Alloc(this);
            solver = s;
            ctx = solver.Context;
            on_clause = onc;
            on_clause_eh = _on_clause;
            Native.Z3_solver_register_on_clause(ctx.nCtx, solver.NativeObject, GCHandle.ToIntPtr(gch), on_clause_eh);
        }

        /// <summary>
        /// Release private memory.
        /// </summary>            
        ~OnClause()
        {
            Dispose();
        }
        
        /// <summary>
        /// Must be called. The object will not be garbage collected automatically even if the context is disposed
        /// </summary>
        public virtual void Dispose()
        {
            if (!gch.IsAllocated)
                return;
            gch.Free();
        }
    }
}

File: src/api/java/OnClause.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    OnClause.java

Abstract:

    Callback on clause inferences

Author:

    Nikolaj Bjorner (nbjorner) 2022-10-19

Notes:

**/

package com.microsoft.z3;

/**
 * Clause inference callback.
 * <p>
 * Allows users to observe clauses learned during solving.
 * Useful for custom learning strategies, clause sharing in parallel solvers,
 * debugging, and proof extraction.
 * </p>
 * <p>
 * Usage: create an instance, override {@link #onClause(Expr, int[], ASTVector)},
 * and close the instance when done.
 * </p>
 **/
public class OnClause implements AutoCloseable {

    private long javainfo;
    private final Context ctx;

    /**
     * Creates an on-clause callback for the given solver.
     *
     * @param ctx    The Z3 context
     * @param solver The solver to register the callback with
     * @throws Z3Exception
     **/
    public OnClause(Context ctx, Solver solver) {
        this.ctx = ctx;
        javainfo = Native.onClauseInit(this, ctx.nCtx(), solver.getNativeObject());
    }

    /**
     * Called when a clause is inferred during solving.
     * <p>
     * The life-time of {@code proof_hint} and {@code literals} is limited to
     * the scope of this callback. If you want to store them, you must duplicate
     * the expressions or extract the literals before returning.
     * </p>
     *
     * @param proof_hint A partial or comprehensive derivation justifying the inference (may be null)
     * @param deps       Dependency indices
     * @param literals   The inferred clause as a vector of literals
     **/
    public void onClause(Expr<?> proof_hint, int[] deps, ASTVector literals) {}

    /**
     * Internal wrapper called from JNI. Do not override.
     **/
    final void onClauseWrapper(long proofHintPtr, int[] deps, long literalsPtr) {
        Expr<?> proof_hint = proofHintPtr != 0 ? (Expr<?>) Expr.create(ctx, proofHintPtr) : null;
        ASTVector literals = new ASTVector(ctx, literalsPtr);
        onClause(proof_hint, deps, literals);
    }

    /**
     * Unregisters the callback and frees associated resources.
     * Must be called when the callback is no longer needed.
     **/
    @Override
    public void close() {
        if (javainfo != 0) {
            Native.onClauseDestroy(javainfo);
            javainfo = 0;
        }
    }
}
